### DISCUSSION TITLE
Support declaring targets and CMake package names in the manifest

### DISCUSSION THREAD

**Discussion by coryan** [microsoft/vcpkg#17863] [Ideas]
First, thank you for implementing manifests, they are awesome.

**Is your feature request related to a problem? Please describe.**

I would like to declare the dependencies for a project in a single place. Today I need to do this in three places:

- The manifest file (`vcpkg.json`) needs to list the package names.
- My `CMakeLists.txt` file needs to have a `find_package(...)` for each package, with a name that may not match the manifest name.
- My `CMakeLists.txt` file needs to add the `target_link_libraries(....)` for each target, which again may not match the package name.

It would be really neat to put all this information in a single place.

**Proposed solution**

Couple of ideas, maybe terrible ones:

- Allow (but do not require) applications to include a couple of fields in the `vcpkg.json` manifest file:

```
{
  "name": "fibo",
  "version-string": "0.1.0",
  "dependencies": [
    "boost-program-options"
    "fmt",
    "range-v3",
    "gtest"
  ],
  "cmake_packages": {
    "Boost": ["program_options"],
    "fmt": [],
    "range-v3", [],
    "gtest": ["GTest::main"],
  },
  "cmake_targets": {
    "fibo": [
        "Boost::program_options",
        "Boost::boost",  
        "fmt::fmt",
        "range-v3::range-v3"
    ],
   "fibo_test": ["fmt::fmt", "GTest::main"]
  ]
}
```

The CMake file would then use these as follows:

```
find_package(VcpkgMagic REQUIRED)
vcpkg_find_packages() # <--- finds all the packages in `cmake_packages` listed above
vcpkg_list_targets("fibo" fibo_targets) # <--- now fibo_targets contains the list in the manifest
```

**Describe alternatives you've considered**

-# Obviously just maintaining the data in `vcpkg.json` and `CMakeLists.txt`.

-# Less magical, but still helpful, keep all the data in one place and use the `vcpkg` tool to print the relevant bits. For example this command:

```sh 
vcpkg manifest --dump cmake_targets
```

could produce the relevant CMake snippet:

```
find_package(Boost REQUIRED COMPONENTS program_options)
find_package(fmt REQUIRED)
find_package(range-v3 REQUIRED)
```

That would at least reduce the maintenance burden (as one would just need to cut&paste the output), and may be automated by generating the snippet and including it into the CMake file.

**Additional context**

N/A

**Comment by Neumann-A** [comment]
`vcpkg.json` and `CMakeLists.txt` needs to be separated to not induce tight coupling between the project and vcpkg. As such your alternative solution is the only viable one. Another alternative would be to search the `_VCPKG_INSTALL_DIR` for all `<packageName>Config.cmake` and do a `find_package()` call for all of them. The problem with that is that needed dependencies which are not installed, due to users mistake in not listing them in the manifests, will not be searched for Furthermore it also introduces coupling. 
So the important part here is that coupling between vcpkg and the project should be avoided under any circumstances to allow other package managers like conan to also work.

**Comment by coryan** [comment]
> `vcpkg.json` and `CMakeLists.txt` needs to be separated to not induce tight coupling between the project and vcpkg.

I understand why many folks would want that.  I am suggesting that some many not, and can accept the tighter coupling for lower maintenance effort.

> So the important part here is that coupling between vcpkg and the project should be avoided under any circumstances to allow other package managers like conan to also work.

For the sake of argument, assume a user does not care if conan works or not. And/or they are willing to do away with standalone builds, or many they do something like:

```
if (DEVELOPMENT_MODE_EASY_TO_USE_BUT_TIGHTLY_COUPLED_TO_VCPKG)
find_package(VcpkgMagic REQUIRED)
vcpkg_find_packages() # <--- finds all the packages in `cmake_packages` listed above
vcpkg_list_targets("fibo" fibo_targets) # <--- now fibo_targets contains the list in the manifest
else ()
# Code here updated only when needed
endif ()
```

Feel free to close with "out of scope" or something.

**Comment by alexsmn** [comment]
> vcpkg.json and CMakeLists.txt needs to be separated to not induce tight coupling between the project and vcpkg. 

Totally agree.

There is the `vcpkg.cmake` toolchain fail that might do all the work. It already installs vcpkg packages when needed. Do you think it can add `find_package` instructions for all dependent packages specified in `vcpkg.json`?

And yeah, manifests are awesome!
